adapters/repos/db/lsmkv/recover_from_wal_integration_test.go

```diff
@@ -444,3 +444,202 @@ func TestSetStrategy_RecoverFromWAL(t *testing.T) {
 		})
 	})
 }
+
+func TestMapStrategy_RecoverFromWAL(t *testing.T) {
+	rand.Seed(time.Now().UnixNano())
+	dirNameOriginal := fmt.Sprintf("./testdata/%d-original", rand.Intn(10000000))
+	dirNameRecovered := fmt.Sprintf("./testdata/%d-recovered", rand.Intn(10000000))
+	os.MkdirAll(dirNameOriginal, 0o777)
+	os.MkdirAll(dirNameRecovered, 0o777)
+	defer func() {
+		err := os.RemoveAll(dirNameOriginal)
+		fmt.Println(err)
+		err = os.RemoveAll(dirNameRecovered)
+		fmt.Println(err)
+	}()
+
+	t.Run("without prior state", func(t *testing.T) {
+		b, err := NewBucket(testCtx(), dirNameOriginal, nullLogger(),
+			WithStrategy(StrategyMapCollection))
+		require.Nil(t, err)
+
+		// so big it effectively never triggers as part of this test
+		b.SetMemtableThreshold(1e9)
+
+		rowKey1 := []byte("test1-key-1")
+		rowKey2 := []byte("test1-key-2")
+
+		t.Run("set original values and verify", func(t *testing.T) {
+			row1Map := []MapPair{
+				{
+					Key:   []byte("row1-key1"),
+					Value: []byte("row1-key1-value1"),
+				}, {
+					Key:   []byte("row1-key2"),
+					Value: []byte("row1-key2-value1"),
+				},
+			}
+
+			row2Map := []MapPair{
+				{
+					Key:   []byte("row2-key1"),
+					Value: []byte("row2-key1-value1"),
+				}, {
+					Key:   []byte("row2-key2"),
+					Value: []byte("row2-key2-value1"),
+				},
+			}
+
+			for _, pair := range row1Map {
+				err = b.MapSet(rowKey1, pair)
+				require.Nil(t, err)
+			}
+
+			for _, pair := range row2Map {
+				err = b.MapSet(rowKey2, pair)
+				require.Nil(t, err)
+			}
+
+			res, err := b.MapList(rowKey1)
+			require.Nil(t, err)
+			assert.Equal(t, row1Map, res)
+			res, err = b.MapList(rowKey2)
+			require.Nil(t, err)
+			assert.Equal(t, res, row2Map)
+		})
+
+		t.Run("replace an existing map key", func(t *testing.T) {
+			err = b.MapSet(rowKey1, MapPair{
+				Key:   []byte("row1-key1"),        // existing key
+				Value: []byte("row1-key1-value2"), // updated value
+			})
+			require.Nil(t, err)
+
+			row1Updated := []MapPair{
+				{
+					Key:   []byte("row1-key1"),
+					Value: []byte("row1-key1-value2"), // <--- updated, rest unchanged
+				}, {
+					Key:   []byte("row1-key2"),
+					Value: []byte("row1-key2-value1"),
+				},
+			}
+
+			row2Unchanged := []MapPair{
+				{
+					Key:   []byte("row2-key1"),
+					Value: []byte("row2-key1-value1"),
+				}, {
+					Key:   []byte("row2-key2"),
+					Value: []byte("row2-key2-value1"),
+				},
+			}
+
+			res, err := b.MapList(rowKey1)
+			require.Nil(t, err)
+			// NOTE: We are accepting that the order is changed here. Given the name
+			// "MapCollection" there should be no expectations regarding the order,
+			// but we have yet to validate if this fits with all of the intended use
+			// cases.
```

Comment:
Good point, interested in seeing user expectations about this in the future

Reply:
Nice catch, this was just due to copy/pasting from an old test. That's actually not true anymore. Starting with BM25 is now stored always-sorted. I'll remove this comment and then merge.